pkg/tempo/promsdprocessor/prom_sd_processor.go

```diff
@@ -0,0 +1,216 @@
+package promsdprocessor
+
+import (
+	"context"
+	"net"
+	"strings"
+	"sync"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/discovery"
+	sd_config "github.com/prometheus/prometheus/discovery/config"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/relabel"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componenterror"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/pdata"
+)
+
+const (
+	addressLabel = "__address__"
+)
+
+type promServiceDiscoProcessor struct {
+	nextConsumer consumer.TraceConsumer
+	discoveryMgr *discovery.Manager
+
+	relabelConfigs map[string][]*relabel.Config
+	hostLabels     map[string]model.LabelSet
+	mtx            sync.Mutex
+
+	logger log.Logger
+}
+
+func newTraceProcessor(nextConsumer consumer.TraceConsumer, scrapeConfigs []*config.ScrapeConfig) (component.TraceProcessor, error) {
+	logger := log.With(util.Logger, "component", "tempo service disco")
+	mgr := discovery.NewManager(context.Background(), logger, discovery.Name("tempo service disco"))
+
+	relabelConfigs := map[string][]*relabel.Config{}
+	cfg := map[string]sd_config.ServiceDiscoveryConfig{}
+	for _, v := range scrapeConfigs {
+		cfg[v.JobName] = v.ServiceDiscoveryConfig
+		relabelConfigs[v.JobName] = v.RelabelConfigs
+	}
+
+	err := mgr.ApplyConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if nextConsumer == nil {
+		return nil, componenterror.ErrNilNextConsumer
+	}
+	return &promServiceDiscoProcessor{
+		nextConsumer:   nextConsumer,
+		discoveryMgr:   mgr,
+		relabelConfigs: relabelConfigs,
+		hostLabels:     make(map[string]model.LabelSet),
+		logger:         util.Logger,
+	}, nil
+}
+
+func (p *promServiceDiscoProcessor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	rss := td.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+		if rs.IsNil() {
+			continue
+		}
+		r := rs.Resource()
+		if r.IsNil() {
+			continue
+		}
+
+		p.processAttributes(r.Attributes())
```

Comment:
Small nit-pick: can you put this file in roughly call order? It'd be easier to read top-down if `processAttributes` was defined right after the method it's first called from.